### PR TITLE
Kubelet monitoring

### PR DESCRIPTION
**Pull request by Q-Lee**
Sidecar to monitor a single kubelet, and push results to the GCM v3 API.

### CODE REVIEW

**Comment by Q-Lee** [comment]
For the life of me, I cannot get "continuous-integration/travis-ci/pr" to pass. Yesterday it was complaining about mungegithub, and today I'm seeing this:

_The command "hack/verify-all.sh -v" exited with 1.
...
The command "hack/for-go-proj.sh test" exited with 0.
Done. Your build exited with 1._

Does anyone know what's up?

**Comment by gmarek** [comment]
```
Boilerplate header is wrong for: /home/travis/gopath/src/k8s.io/contrib/kubelet-to-gcm/monitor/kubelet_to_gcm.go
Boilerplate header is wrong for: /home/travis/gopath/src/k8s.io/contrib/kubelet-to-gcm/monitor/kubelet_to_gcm_test.go
Found a flag declared with an _ but which is not explicitly listed as a valid flag name in hack/verify-flags/excluded-flags.txt
Are you certain this flag should not have been declared with an - instead?
use_test
```

**Comment by Q-Lee** [comment]
Ah, I'm blind. That would explain travis failing. I had a copy/paste error with the license header. I've renamed the flag, and that should help too.

**Comment by gmarek** [comment]
@Q-Lee - squash please.

**Comment by Q-Lee** [comment]
Poke. Does anyone have time to do this CR, or should I search someone else out?

**Comment by gmarek** [comment]
@Q-Lee I'm leaving on vacation for a week, so I won't be able to review this. Sorry.

**Comment by mwielgus** [comment]
Could you please split this PR into 2 - one with godeps and one with the real code? Github makes reviewing 600 files super inconvenient.

**Comment by Q-Lee** [comment]
How about 2 commits in the same PR, so you can view one with the code, and one with the godeps?

**Comment by mwielgus** [comment]
Sorry, that was what I meant :).

**Comment by mwielgus** [comment]
Some extra commits got into this pr.

**Comment by Q-Lee** [comment]
Ok, I forgot to change the flags file when I was restructuring commits. I also did a new rebase, but now I've picked up a different 2 commits....

**Comment by mwielgus** [comment]
Still there are 2 commits that should not be in this pr.

**Comment by Q-Lee** [comment]
Bump.

File: kubelet-to-gcm/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# addon-resizer
```
Comment (fgrzadkowski) [inline kubelet-to-gcm/README.md]: I must be missing something. This file is about addon-resize, while the rest of the code looks unrelated. Is it a mistake?

**Comment by fgrzadkowski** [comment]
@Q-Lee Can you please give more context about this change? What is this code/daemon supposed to do?

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+import (
+	"fmt"
+	"time"
+
+	v3 "google.golang.org/api/monitoring/v3"
+	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/stats"
+)
+
+var (
+	reservedCoresMD = &metricMetadata{
```
Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: What is the difference between this metrics and metrics exposed by kubelet in summary api?

**Comment by mwielgus** [comment]
This code seems to be equivalent to heapster getting data from a single kubelet. It might be more beneficial to actually modify heapster to handle this use case, not to create another entity that duplicates the job.

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 2 above)
Comment (Q-Lee) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: These are the GCM/stack driver metrics. I'm pushing master component metrics to stack driver from GKE's hosted masters.

File: kubelet-to-gcm/README.md (hunk 1 above)
Comment (Q-Lee) [inline kubelet-to-gcm/README.md]: Yes. Let me update the README, which will also give context on what this code/daemon is supposed to do.

**Comment by Q-Lee** [comment]
I originally did start on the track of modifying heapster to hit a single kubelet (w/no apiserver). But it violated quite a few assumptions in the heapster code, so it didn't feel right to restructure heapster for the weird GKE master.

File: kubelet-to-gcm/README.md (hunk 1 above)
Comment (fgrzadkowski) [inline kubelet-to-gcm/README.md]: I think this is still not fixed. This file describes addon resizer, not kubelet-to-gcm.

**Comment by fgrzadkowski** [comment]
As discussed with @mwielgus he will take a look at this PR.

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+	}
+	diskUsedMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "container.googleapis.com/container/disk/bytes_used",
+	}
+	memTotalMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "container.googleapis.com/container/memory/bytes_total",
```
Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: Are theses names in sync with heapster metrics?

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+		for _, container := range pod.Containers {
+			containerName := container.Name
+			// Check for duplicates
+			if container.StartTime.Time.Before(metricsSeen[containerName]) || container.StartTime.Time.Equal(metricsSeen[containerName]) {
+				continue
+			}
+			metricsSeen[containerName] = container.StartTime.Time
+			var containerSeries []*v3.TimeSeries
+
+			monitoredLabels := map[string]string{
```
Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: I believe that Heapster pushes a bit more metrics (like image name). Are the names in sync with heapster?

File: kubelet-to-gcm/monitor/main/daemon.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+const (
+	missingValue = "/MISSING"
+	scope        = "https://www.googleapis.com/auth/monitoring.write"
+	testPath     = "https://test-monitoring.sandbox.googleapis.com"
+)
+
+var (
+	// Flags to identify the Kubelet.
+	zone    = flag.String("zone", "us-central1-b", "The zone where this kubelet lives.")
```
Comment (mwielgus) [inline kubelet-to-gcm/monitor/main/daemon.go]: Why do you need this?

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 3 above)
Comment (Q-Lee) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: It's similar, but the heapster metrics for GCM are custom metrics. We're pushing the v3 metrics here.

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 4 above)
Comment (Q-Lee) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: This doesn't take 100% of heapster metrics: only the ones used by GCM pre-defined metrics. These metrics don't exist in heapster atm (heapster pushes custom metrics to the v2 API).

File: kubelet-to-gcm/monitor/main/daemon.go (hunk 5 above)
Comment (Q-Lee) [inline kubelet-to-gcm/monitor/main/daemon.go]: As opposed to? I suppose I could try taking it from the metadata server on the host. Is that what you're suggesting?

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 4 above)
Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: Ok

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 3 above)
Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: Ok.

Comment (mwielgus) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: Why do we want to have 2 different falvors of metrics?

File: kubelet-to-gcm/monitor/main/daemon.go (hunk 5 above)
Comment (mwielgus) [inline kubelet-to-gcm/monitor/main/daemon.go]: Yes, Heapster, for instance doesn't need any zone/project data to be passed to push data to GCM.

Comment (Q-Lee) [inline kubelet-to-gcm/monitor/main/daemon.go]: Using the metadata server isn't an option when you're hitting a remote kubelet. Since I'm no longer doing that, I suppose I could hit get it programmatically.

File: kubelet-to-gcm/monitor/kubelet_to_gcm.go (hunk 3 above)
Comment (Q-Lee) [inline kubelet-to-gcm/monitor/kubelet_to_gcm.go]: The v2 custom metrics defined in heapster are, afaik, dead code. GKE actually uses a different metric schema altogether. When heapster supports the v3 API, it should use the schema GKE uses, and the custom metrics it uses now.

**Comment by mwielgus** [comment]
LGTM, please rebase.
